weasyl/two_factor_auth.py

```diff
@@ -0,0 +1,383 @@
+    tfa_response = tfa_response.replace(' ', '')
+    if len(tfa_response) == LENGTH_TOTP_CODE:
+        # Retrieve the encrypted 2FA secret from `userid`'s login record
+        tfa_secret_encrypted = d.engine.scalar("""
+            SELECT twofa_secret
+            FROM login
+            WHERE userid = %(userid)s
+        """, userid=userid)
+        # Decrypt the 2FA secret to be usable by pyotp
+        tfa_secret = _decrypt_totp_secret(tfa_secret_encrypted)
+        # Validate supplied 2FA response versus calculated current TOTP value.
+        totp = pyotp.TOTP(tfa_secret)
+        # Return the response of the TOTP verification; True/False
+        return totp.verify(tfa_response)
+    elif len(tfa_response) == LENGTH_RECOVERY_CODE:
+        # Check if `tfa_response` is valid recovery code; consume according to `consume_recovery_code`,
+        #  and return True if valid, False otherwise
+        return is_recovery_code_valid(userid, tfa_response, consume_recovery_code)
+    # Other lengths are invalid; return False
+    else:
+        return False
+
+
+def get_number_of_recovery_codes(userid):
+    """
+    Get and return the number of remaining recovery codes for `userid`.
+
+    Parameters:
+        userid: The userid for which to check the count of recovery codes.
+
+    Returns:
+        An integer representing the number of remaining recovery codes.
+    """
+    return d.engine.scalar("""
+        SELECT COUNT(*)
+        FROM twofa_recovery_codes
+        WHERE userid = %(userid)s
+    """, userid=userid)
+
+
+def generate_recovery_codes():
+    """
+    Generate a set of valid recovery codes.
+
+    Character set is defined by `libweasyl.security`, (ASCII+Numbers), limited to uppercase via
+    .upper() for readability.
+
+    Parameters: None
+
+    Returns: A set of length `_TFA_RECOVERY_CODES` where each code is `LENGTH_RECOVERY_CODE`
+    characters in length.
+    """
+    return {security.generate_key(LENGTH_RECOVERY_CODE).upper() for i in range(_TFA_RECOVERY_CODES)}
+
+
+def store_recovery_codes(userid, recovery_codes):
+    """
+    Store generated recovery codes into the recovery code table, checking for validity.
+
+    Parameters:
+        userid: The userid to save the recovery codes for.
+        recovery_codes: Comma separated unicode string of recovery codes.
+
+    Returns: Boolean True if the codes were successfully saved to the database, otherwise
+    Boolean False
+    """
+    # Force the incoming string to uppercase, then split into a list
+    codes = recovery_codes.upper().split(',')
+    # The list must exist and be equal to the current codes to generate
+    if len(codes) != _TFA_RECOVERY_CODES:
+        return False
+    # Make sure all codes are `LENGTH_RECOVERY_CODE` characters long, as expected
+    for code in codes:
+        if len(code) != LENGTH_RECOVERY_CODE:
+            return False
+
+    # Store the recovery codes securely by hashing them with bcrypt (as login.passhash())
```

Comment:
For when the inevitable code-review comments come down: s/ (as login.passhash())// -- Just noticed this fragment remained from when I opted to use the slightly reduced work-factor.